### ISSUE TITLE
LinearDensity Invalid Box Problem

### ISSUE THREAD

**Issue by criosx** [MDAnalysis/mdanalysis#2690]
When attempting to calculate the linear density of a molecular selection using MDAnalysis.analysis.lineardensity.LinearDensity(), and the selection does not have a dimension (box) set, i.e. [0, 0, 0, 90, 90, 90], the routine raise a value error:
```
density = MDAnalysis.analysis.lineardensity.LinearDensity(selection, grouping='atoms')
density.run()
```
The error is raised in the function wrap() in groups.py, line 1422:
```
        if not np.all(box > 0.0) or box.shape != (6,):
            raise ValueError("Invalid box: Box has invalid shape or not all "
                             "box dimensions are positive. You can specify a "
                             "valid box using the 'box' argument.")
```
Setting the dimensions of the selection to a valid value like in the following example should solve the problem, but still fails with the same error message
```
selection.dimensions=numpy.array([200, 200, 200, 90, 90, 90])
density = lin.LinearDensity(selection grouping='atoms')
density.run()
```
It turns out that in run(), the variable selection.dimensions (at this point, self._ags[0].dimensions) is overwritten by selection.universe.trajectory.dimensions, which cannot be externally set. Trying to call the following will raise an "AttributeError: can't set attribute.".

`selection.universe.trajectory.dimensions=numpy.array([200, 200, 200, 90, 90, 90])`

A make-shift fix to the function run() in base.py is appended (see # Modification). Is there a better way to deal with this? What do I miss? What does “You can specify a valid box using the 'box' argument.” in the original value error message mean? Box argument of which object?

```
    def run(self, start=None, stop=None, step=None, verbose=None):
        """Perform the calculation

        Parameters
        ----------
        start : int, optional
            start frame of analysis
        stop : int, optional
            stop frame of analysis
        step : int, optional
            number of frames to skip between each analysed frame
        verbose : bool, optional
            Turn on verbosity
        """
        logger.info("Choosing frames to analyze")
        # if verbose unchanged, use class default
        verbose = getattr(self, '_verbose', False) if verbose is None else verbose
        
        self._setup_frames(self._trajectory, start, stop, step)
        logger.info("Starting preparation")
        self._prepare()

        temp=self._ags[0].dimensions          # Modification

        for i, ts in enumerate(
                self._trajectory[self.start:self.stop:self.step]):
            
            self._ags[0].dimensions=temp         # Modification

            self._frame_index = i
            self._ts = ts
            # logger.info("--> Doing frame {} of {}".format(i+1, self.n_frames))

            self._single_frame()
            self._pm.echo(self._frame_index)
        logger.info("Finishing up")
        self._conclude()
        return self
```
Thank you very much.

**Comment by lilyminium** [comment]
Hi @criosx, box dimensions are considered a dynamic attribute that updates each frame, because the box will change in NPT simulations. The easiest solution is probably to set the box for each timestep/frame with [on-the-fly transformations](https://www.mdanalysis.org/docs/documentation_pages/trajectory_transformations.html).

```python
def boxify(ts):
    ts.dimensions = [200, 200, 200, 90, 90, 90]
    return ts

selection.universe.trajectory.add_transformations(boxify)
```

You could also [write your trajectory](https://www.mdanalysis.org/UserGuide/reading_and_writing.html#frames-and-trajectories) to a format with box information and load it back in.
```python
with mda.Writer('my_box_traj.xtc', len(universe.atoms)) as w:
    for ts in universe.trajectory:
        ts.dimensions = [200, 200, 200, 90, 90, 90]
        w.write(universe.atoms)

box_universe = mda.Universe('my_top_file.pdb', 'my_box_traj.xtc')
```


> What does “You can specify a valid box using the 'box' argument.” in the original value error message mean? Box argument of which object?

> The error is raised in the function wrap() in groups.py, line 1422

This would be the `box` argument of the `AtomGroup.wrap()` function, which is called by the `LinearDensity` code. While you can manually pass in a `box` when you yourself call `wrap()`, you would have to alter the `LinearDensity` code for it to work there.
